Searching: getElement*, querySelector* (#254)
javascript-tutorial/it.javascript.info: 2-ui/1-document/04-searching-elements-dom/article.md

## document.getElementById o semplicemente id

Se un elemento ha un attributo `id` possiamo trovarlo, ovunque esso sia, utilizzando il metodo `document.getElementById(id)`

Ad esempio:

```html run
<div id="elem">
  <div id="elem-content">Element</div>
</div>

<script>
  // trova l'elemento
*!*
  let elem = document.getElementById('elem');
*/!*

  // rendi il suo background rosso
  elem.style.background = 'red';
</script>
```

Inoltre, c'è una variabile globale chiamata come l'`id` che fa riferimento all'elemento:

```html run
<div id="*!*elem*/!*">
  <div id="*!*elem-content*/!*">Element</div>
</div>

<script>
  // elem e' un riferimento all'elemento del DOM con id="elem"
  elem.style.background = 'red';

  // id="elem-content" contiene un trattino, quindi non può essere il nome di una variabile
  // ...ma possiamo accedervi tramite parentesi quadre: window['elem-content']
</script>
```

...Questo a meno di dichiarare una variabile JavaScript con lo stesso nome: quest'ultima avrebbe la precedenza:

```html run untrusted height=0
<div id="elem"></div>

<script>
  let elem = 5; // ora elem è 5, non un riferimento a <div id="elem">

  alert(elem); // 5
</script>
```

> ```warn header="Non utilizzate le variabili globali per accedere agli elementi"

Reviewer: Qui non è troppo chiaro cosa si intenda per variabili globali. Quello che intende dire in lingua inglese è che se tu dichiari un elemento con id, verrà automaticamente creata una variabile, e potrai accedere al nodo con `window.[nomeId]`... però non saprei nemmeno io come tradurre in maniera elegante questo titoletto

```suggestion
```warn header="Non utilizzate le variabili globali denominate dall'id per accedere agli elementi"
```

Author: ```suggestion
```warn header="Non utilizzate le variabili-id globali per accedere agli elementi"
```
Che ne dici di cosi'? Un po' banale forse ma e' leggermente piu' chiaro, non trovi?

Reviewer: Si molto piu chiaro cosi! Vai pure con la tua modifica, poi è pronta per il merge

> Questo comportamento è descritto [nella specifica](http://www.whatwg.org/specs/web-apps/current-work/#dom-window-nameditem), quindi è quasi uno standart, ma è supportato principalmente per una questione di compatibilità.

Reviewer: ```suggestion
Questo comportamento è descritto [nella specifica](http://www.whatwg.org/specs/web-apps/current-work/#dom-window-nameditem), quindi è quasi uno standard, ma è supportato principalmente per una questione di compatibilità.
```

</script>
```

Questo metodo è molto potente poiché può utilizzare qualsiasi CSS selector.

```smart header="Può utilizzare anche pseudo-classi"

> Le pseudo-classi come `:hover` e `:active` sono altresì supportate. Ad esempio, `document.querySelectorAll(':hover')` ritornerà una collezione con gli elementi sopra cui punta il cursore in questo momento (seguendo l'ordine di annidamento, dal più esterno `<html>` all'elemento più annidato).

Author: Sono d'accordo.

  alert(chapter.closest('.book')); // UL
  alert(chapter.closest('.contents')); // DIV

  alert(chapter.closest('h1')); // null (perché h1 non è un antenato)
</script>
```

## getElementsBy*

Ci sono altri metodi per cercare nodi tramite tag, classi ecc.

Oggi sono più che altro storia, essendo `querySelector` più potente e più breve da scrivere.

> Qui li studiamo per completezza, mentre possono essere ancora trovati nei vecchi script.

Reviewer: ```suggestion
Qui li studiamo per completezza, poiché possono essere ancora trovati nei vecchi script.
```

</form>

<script>
  // cerca tramite il nome dell'attributo
  let form = document.getElementsByName('my-form')[0];

  // trova tramite la classe contenuta in form
  let articles = form.getElementsByClassName('article');
  alert(articles.length); // 2, trovati due elementi con la classe "article"
</script>
```

## Collections vive

> Tutti i metodi `"getElementsBy*"` ritornano una collection *viva*. Questa collction rispecchia sempre lo stato corrente del documento e si "auto-aggiorna" quando questo cambia.

Reviewer: ```suggestion
Tutti i metodi `"getElementsBy*"` ritornano una collection *viva*. Questa collection rispecchia sempre lo stato corrente del documento e si "auto-aggiorna" quando questo cambia.
```

</tr>
<tr>
<td><code>getElementsByTagName</code></td>
<td>tag o <code>'*'</code></td>
<td>✔</td>
<td>✔</td>
</tr>
<tr>
<td><code>getElementsByClassName</code></td>
<td>classe</td>
<td>✔</td>
<td>✔</td>
</tr>
</tbody>
</table>

> I più utilizzati sono `querySelector` e `querySelectorAll`, ma `getElement(s)By*` può essere utile qualche volta, e si trova nei vecchi script.

Reviewer: ```suggestion
I più utilizzati sono `querySelector` e `querySelectorAll`, ma `getElement(s)By*` può essere utile in alcuni casi, e si trova nei vecchi script.
```
